docker-app/qfieldcloud/core/utils2/storage.py

```diff
@@ -268,32 +323,41 @@ def upload_project_file(
     return key
 
 
-def delete_project_files(project_id: str) -> None:
-    bucket = qfieldcloud.core.utils.get_s3_bucket()
+def delete_all_project_files_permanently(project_id: str) -> None:
     prefix = f"projects/{project_id}/"
-    bucket.object_versions.filter(Prefix=prefix).delete()
+
+    if not not re.match(r"^projects/[\w-]+/.+$", prefix):
+        raise RuntimeError(
+            f"Suspicious S3 deletion of all project files with {prefix=}"
+        )
+
+    delete_by_prefix_versioned(prefix)
 
 
-def delete_file(project: "Project", filename: str):  # noqa: F821
+def delete_project_file_permanently(project: "Project", filename: str):  # noqa: F821
     file = qfieldcloud.core.utils.get_project_file_with_versions(project.id, filename)
 
     if not file:
-        raise Exception("No file with such name in the given project found")
+        raise Exception(
+            f"No file with such name in the given project found {filename=}"
+        )
 
-    file.delete()
+    with transaction.atomic():
+        if qfieldcloud.core.utils.is_qgis_project_file(filename):
+            project.project_filename = None
+            project.save(recompute_storage=True, update_fields=["project_filename"])
 
-    if qfieldcloud.core.utils.is_qgis_project_file(filename):
-        project.project_filename = None
-        project.save(recompute_storage=True)
+        # NOTE auditing the file deletion in the transation might be costly, but guarantees the audit
+        audit(
+            project,
+            LogEntry.Action.DELETE,
+            changes={f"{filename} ALL": [file.latest.e_tag, None]},
+        )
 
-    audit(
-        project,
-        LogEntry.Action.DELETE,
-        changes={f"{filename} ALL": [file.latest.e_tag, None]},
-    )
+        delete_by_prefix_versioned(filename)
 
 
-def delete_file_version(
+def delete_project_file_version_permanently(
```

Comment:
While being paranoid, we could prevent deletion of the last remaining version of the file using this function ? (meaning you'd have to use `delete_project_file_permanently` to actually delete the file). Basically replace `is_deleting_all_versions = True` below by `raise Exception("Cannot delete latest version, it would delete the file !")`

Imagine a case where two users delete a different version of the file more or less at the same time, each one thinking "all fine, I still have the other version left".

Reply:
Ok, convinced, see my last commit.